### DISCUSSION TITLE
Built-in Lagrange Local Node Ordering

### DISCUSSION THREAD

**Discussion by NamjaeChoi** [CEED/libCEED#1689] [Q&A] [answered]
Hello,

Where can I find the local node ordering of each element type used by the built-in Lagrange basis? We want to check if it's consistent with the one currently used in MOOSE.

Thank you.

**Comment by jeremylt** [comment, marked as answer]
For a 2D cubic element, the ordering is
```
01 02 03 04
05 06 07 08
09 10 11 12
13 14 15 16
```

**Reply by jedbrown** [reply]
And for MOOSE, you already have bases (tabulated by libMesh) that are probably better to use than libCEED's bulit-in constructors (which exist for demonstration purposes, but have no further advantages). I guess libMesh may lose the tensor product structure if you use the direct tabulation. If it is not a tensor product ordering, you can permute the libMesh DoF map when creating the element restriction.

**Reply by NamjaeChoi** [reply]
Yeah that is our basic approach and works when QPs are fixed to the reference location. But we have cases where we should evaluate shape functions on arbitrary QPs and we don't see a good way to handle that.

**Reply by jeremylt** [reply]
Are you referring to something where each element on the mesh has its own quadrature space with points at arbitrary locations? Can you provide an example?

**Reply by NamjaeChoi** [reply]
In MOOSE DG kernels, the face quadrature values of neighbor is computed on the face QPs of the current element.

**Reply by jeremylt** [reply]
I guess I'm not seeing how using the built in libCEED H1 Lagrange basis constructor would help with that?

**Reply by NamjaeChoi** [reply]
Because libCEED seems to have `CeedBasisApplyAtPoints` that can evaluate arbitrary QP values with the built-in Lagrange?

**Reply by jeremylt** [reply]
You do not need to use the built in H1 Lagrange to use `CeedBasisApplyAtPoints`, any basis with a tensor product structure is fine.

Note - `CeedBasisApplyAtPoints` is much slower than using a typical quadrature space, especially for GPUs. If you can create a basis object that represents mapping into the other quadrature space, that will give you much better performance.

**Reply by NamjaeChoi** [reply]
@lindsayad Not sure if I understood. Would you like to hop in?

**Reply by jeremylt** [reply]
To throw some rough numbers at it, for the action of a gradient it's 25-30x slower to use AtPoints instead of a standard basis action.

**Reply by lindsayad** [reply]
Let's say we have a two-point quadrature rule on a 1D face of a 2D quad or tri. We'll call the points 0 and 1. For a DG calculation we need to be able to evaluate the shape functions on the neighboring element at points 0 and 1. A default quadrature rule that didn't use the information from the "primary" element would create its own points 0' and 1'. In general 0 may correspond to 0' or 1', so we cannot use a general qudrature rule on the neighboring element side; we must use the info from the primary element. So the way we do it in MOOSE is that we inverse map the points 0 and 1 in physical space into the reference space of the neighboring element to correctly determine 0' and 1' and guarantee that 0 corresponds to 0' and 1 corresponds to 1'. Ideas for guaranteeing this correspondence in a more GPU performant way are very welcome

**Reply by knepley** [reply]
There are a number of questions here:
1) Why would one cell care what quadrature another cell is using? The integrals are not connected, even if the quadrature points are pushed to the boundary. You usually care about dofs matching up between cells, but not quad points.
2) Are you worried about about face orientation in the cell? This matter for shape function evaluation, but you handle this when you make your numbering for the local-to-cell map.

**Reply by jrwrigh** [reply]
_Disclaimer: I'm not intimately familiar with DG methods, so I my suggestion may not make sense._

If we were having to use libCEED to write a code that evaluates both a volume integral and a flux integral between elements, I have those integrals evaluated in CeedOperator and combine them into a composite operator. 

The primary reason why is that a CeedOperator makes to guarantees about what quadrature points are being passed to a QFunction, whether they're in the same element, or what location within an element they are at. In other words, there's no ability to distinguish a volume quadrature point from a boundary quadrature point. So we have to make two different CeedOperators; one for volume quadrature points, another for volume quadrature points.

To handle the boundary flux CeedOperator, I think it'd need 2 inputs, one for the "left" element, one for the "right" element. The CeedBasis would be an interpolation from those element DoFs to the face quadrature points. Tricky part there becomes orientation of DoFs. One way to address that is to have a CeedBasis per orientation and thus a CeedOperator per relative orientation of the "left" and "right" element faces. HONEE and Ratel do a similar operator for domain boundary integrals, so it's not untrodden territory. I think Matt's idea of a different element restriction could also work.

Note the above strategy would require evaluating the solution at quadrature points twice if the volume and boundary quadrature points are collocated. I'm not sure if there's a way to avoid that, or if it's worth trying to avoid it.

**Reply by jedbrown** [reply]
Is this for hanging node or fully-non-conforming meshes. If so, I think using `*AtPoints` makes good sense for those non-conforming faces. The number of surface quadrature points will be small compared to the volume. For conforming faces, you can use a regular quadrature. You probably have more conforming faces in the global mesh than non-conforming faces.

**Reply by lindsayad** [reply]
> Why would one cell care what quadrature another cell is using?

One thing we do in MOOSE is that if one element is p-refined then we typically upgrade it's quadrature rule order. In such a case, if a neighboring element is less p-refined, then we elect to use the higher order quadrature on the face corresponding to the more p-refined element

> You usually care about dofs matching up between cells, but not quad points

I don't understand this statement. How do you correctly calculate a jump term if the left and right evaluations are not collocated? E.g. $\int_L \psi_L (u_L - u_R) dS = \sum_{qp = 1}^{n_{qp}} \psi_{L,qp} (u_{L,qp} - u_{R,qp}) J_{qp} w_{qp}$

> Are you worried about about face orientation in the cell? 

Yes

**Reply by knepley** [reply]
Matching those jump terms is about _dofs_, not about quadrature points. In fact, you could easily evaluate the left and right integrals with different quadratures.

**Reply by jedbrown** [reply]
@knepley That doesn't make sense to me. It's DG so there is no dof continuity or matching of dofs between elements.

@lindsayad I think `AtPoints` is an acceptable way to do this. An alternative would be to classify faces into batches based on their pairing (surely there are a limited number of combinations if you're only p-nonconforming) and create separate `CeedOperator`s for each -- these can be bundled into a composite operator so the batching would be local. If you choose `AtPoints`, I would still encourage to make a batch with all the conforming faces (which would be most of them) so the efficiency hit is only incurred for the nonconforming faces.

**Reply by knepley** [reply]
@jedbrown Matching in the sense that you have to know which dofs you are integrating from each side. Are you saying that you have to match quadrature points? I don't see how that is true. The integrals can be done independently.

**Reply by jedbrown** [reply]
The integral is of a jump term, tested from each side. The dofs on each side are entirely decoupled except for the jump term. No matter how you shake it, you need to evaluate basis functions from the left element to quadrature points for an integral with test functions from the right element (and vice-versa). What are you proposing?

**Reply by knepley** [reply]
Oh, this is a face integral. Why was Alex talking about the quadrature order in the cell? That was terribly confusing.

If you are doing a face integral, then all that matters is the face quadrature, and yes the cell on each side has to evaluate at the face quadrature. It does not have to be DG. We do this in PyLith too.

**Reply by jedbrown** [reply]
If it's a face quadrature and the faces are non-conforming, the quadrature on that face won't be "normal" for one of the elements that need to evaluate to it. Using libCEED's `AtPoints` evaluation for the nonconforming face integrals would be one way to do that without any assumed structure or need for manual batching of elements.

**Reply by knepley** [reply]
Are these really hanging node constraints? Doesn't that mean you don't just want to evaluate the little guys on the other side, but you want to implement the constraint as well before you evaluate?

**Reply by jedbrown** [reply]
DG doesn't have hanging node constrains because you're solving for a discontinuous solution. You only need to evaluate to quadrature points on the interface. If it's a 2:1 situation, you perform integration on the small faces.

**Reply by knepley** [reply]
Yeah, if all you have to do is integrate on that face, using AtPoints for all non-conforming faces is the easiest thing.

**Reply by jeremylt** [reply]
Though, my point above I think is still important. AtPoints is significantly slower than standard basis evaluations, so I'd only use it when you cannot do otherwise

**Reply by lindsayad** [reply]
Sorry @knepley that face vs. volume was unclear. Yes I am talking about the face integrals. And thanks to @jedbrown it's clear how we might "optimally" handle the non-conforming case with `AtPoints`. However

> For conforming faces, you can use a regular quadrature.

It is still unclear to me how to implement this more basic/common case. I believe there have been two suggestions? One was @jrwrigh:

> To handle the boundary flux CeedOperator, I think it'd need 2 inputs, one for the "left" element, one for the "right" element. The CeedBasis would be an interpolation from those element DoFs to the face quadrature points. Tricky part there becomes orientation of DoFs. One way to address that is to have a CeedBasis per orientation and thus a CeedOperator per relative orientation of the "left" and "right" element faces.

And then the other suggestion is two different element restrictions?

**Reply by knepley** [reply]
Oh, I think you handle this the same way we do FV for RDyCore. Namely, you have a separate restriction for the left and right cells, and that restriction has the orientation embedded in it. This does not resolve the "flip" for some elements, so it would be nice if the CEED interface would do that.

**Reply by jedbrown** [reply]
Yeah, restrict from the left cell and use a Basis to evaluate to quadrature points, and the same from the right. These will be two inputs to the qfunction. In RDycore (FV), the qfunction sets two outputs (for contribution to each cell) that differ by a sign (and cell volumes in that case).

**Reply by lindsayad** [reply]
Awesome, thanks guys. It's pretty clear now
